### PR TITLE
Prohibit null key on HMAC.

### LINKED ISSUE #2002: HMACXxx is throwing NRE when passing a null key
When instanciating a derived class from `HMAC` like `HMACSHA256`, there is not validation of the `byte[]` provided as key. This is causing a `NullReferenceException`.
```
new System.Security.Cryptography.HMACSHA256(null);
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Internal.Cryptography.HMACCommon.ChangeKey(Byte[] key)
   at System.Security.Cryptography.HMACSHA256..ctor(Byte[] key)
```

Passing a null key is probably not a good practice 🙄  , but the exception may be more explicit.

### PR DESCRIPTION
**Pull request by vcsjones**
Fixes #2002 

3.x and 2.1, and .NET Framework currently throw a NRE with a null HMAC key. dotnet/corefx/pull/42567 made changes that removed the NRE and caused a null byte[] to be converted to an empty span, thus an empty key is used. This change restores the throwing behavior of using a null key, using a better exception.

/cc @bartonjs

### CODE REVIEW

File: src/libraries/System.Security.Cryptography.Algorithms/tests/HmacMD5Tests.cs [hunk 1]
```diff
@@ -85,5 +85,11 @@ public void HMacMD5_Rfc2104_2()
         {
             VerifyHmacRfc2104_2();
         }
+
+        [Fact]
+        public void HMacMD5_ThrowsArgumentNullForNullConstructorKey()
+        {
+            Assert.Throws<ArgumentNullException>(() => new HMACMD5(null));
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Algorithms/tests/HmacMD5Tests.cs]:
```suggestion
            AssertExtensions.Throws<ArgumentNullException>("key", () => new HMACMD5(null));
```

(and similarly for the rest)

**Comment by bartonjs** [comment]
Thanks for jumping on it quickly! Looks good once the exception tests are updated to make sure that we're assigning the right parameter names.

**Review by GrabYourPitchforks — APPROVED** [review]

File: src/libraries/System.Security.Cryptography.Algorithms/tests/HmacMD5Tests.cs (hunk 1 above)
Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.Algorithms/tests/HmacMD5Tests.cs]: Also, while it's slightly redundant, it's probably worth verifying the set_Key exception on each implementation type.

**Comment by vcsjones** [comment]
@bartonjs that's a neat little assertion helper. Updated.

File: src/libraries/System.Security.Cryptography.Algorithms/tests/HmacMD5Tests.cs (hunk 1 above)
Comment (vcsjones) [reply src/libraries/System.Security.Cryptography.Algorithms/tests/HmacMD5Tests.cs]: > Also, while it's slightly redundant, it's probably worth verifying the set_Key exception on each implementation type.

Not sure I follow - are you suggesting making the same `AssertExtensions` change to to `Key` setter's test?

Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.Algorithms/tests/HmacMD5Tests.cs]: Apparently I had a brain-o.  I read the last test as using `HMAC.Create()`, not a test-specific `Create()` from inheritance, so I thought we were only testing one algorithm for set_Key throwing.

But HMAC.Create() throws PNSE, so it clearly isn't... and I've now reconstructed what that test does.

::whistles innocently::

**Review by bartonjs — APPROVED** [review]

**Review by stephentoub — APPROVED** [review]

**Pull request merged**